### PR TITLE
CH-352: Added support for skip, sort, limit and filter.

### PR DESCRIPTION
**Pull request by akash2144**
Signed-off-by: Akash Dudhane <akash.a.dudhane@seagate.com>

# Problem Statement
- [CH-352](https://jts.seagate.com/browse/CH-352)
Alert data store adding support of advance options

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added - Full test details - https://jts.seagate.com/secure/attachment/533495/CH-352.txt
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [x] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [x] Changes done to WIKI / Confluence page / Quick Start Guide  https://seagate-systems.atlassian.net/wiki/spaces/CH/pages/1074757633/LLD+Event+workflow+and+management+backend#Data-Store-Backend-(DSB)

# Unit test results
Note : One provisioner test failed which is not related to this PR.
```
[root@cortx-ha-headless-svc cortx-halo]# python3 test/py/main.py -m 'unit'
==================================================== test session starts ====================================================
platform linux -- Python 3.6.8, pytest-7.0.1, pluggy-1.0.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /cortx-halo, configfile: pytest.ini
collected 27 items / 12 deselected / 15 selected

common/test/test_alert_data_store.py::test_save_data PASSED                                                           [  6%]
common/test/test_alert_data_store.py::test_get_data PASSED                                                            [ 13%]
common/test/test_alert_data_store.py::test_delete_data PASSED                                                         [ 20%]
common/test/test_alert_data_store.py::test_delete_collection PASSED                                                   [ 26%]
common/test/test_datastore_backend.py::test_create_time_series_data_store PASSED                                      [ 33%]
common/test/test_datastore_backend.py::test_create_index PASSED                                                       [ 40%]
common/test/test_datastore_backend.py::test_save_data PASSED                                                          [ 46%]
common/test/test_datastore_backend.py::test_get_data PASSED                                                           [ 53%]
common/test/test_datastore_backend.py::test_delete_index PASSED                                                       [ 60%]
common/test/test_datastore_backend.py::test_delete_collection PASSED                                                  [ 66%]
common/test/test_user_data_store.py::test_save_data PASSED                                                            [ 73%]
common/test/test_user_data_store.py::test_get_data PASSED                                                             [ 80%]
common/test/test_user_data_store.py::test_delete_data PASSED                                                          [ 86%]
common/test/test_user_data_store.py::test_delete_collection PASSED                                                    [ 93%]
provisioner/test/unit/test_network_config.py::test_network_config FAILED                                              [100%]

========================================================= FAILURES ==========================================================
____________________________________________________ test_network_config ____________________________________________________

    def test_network_config():
        config(cfgfile='test_network_config_input.yaml')
>       with open("test_network_config_output.yaml", "r") as f:
E       FileNotFoundError: [Errno 2] No such file or directory: 'test_network_config_output.yaml'

provisioner/test/unit/test_network_config.py:34: FileNotFoundError
--------------------------------------------------- Captured stdout call ----------------------------------------------------
[Errno 2] No such file or directory: '/cortx-halo/provisioner/components/network/test_network_config_input.yaml'
================================================== short test summary info ==================================================
FAILED provisioner/test/unit/test_network_config.py::test_network_config - FileNotFoundError: [Errno 2] No such file or di...
======================================== 1 failed, 14 passed, 12 deselected in 3.84s ========================================
[root@cortx-ha-headless-svc cortx-halo]#
```

### CODE REVIEW

File: common/db_store/db.py [hunk 1]
```diff
@@ -90,26 +90,66 @@ def get_data(self, data_store_name: str, **kwargs):
...
+            skip (int): To skip the first x records of the returned results.
+            count(bool): True / False. To get the total count of documents
+                irrespective of limit & skip field.
 
         Raises:
             DBError: Unable to fetch data.
 
         Returns:
-            list: List of records.
+            tuple: (List of records, Total count of documents)
```
Comment (mukhtar-inamdar) [inline common/db_store/db.py:108]: I think it better to return the first number then a List of records.
call is your.

File: common/db_store/db.py [hunk 2]
```diff
@@ -90,26 +90,66 @@ def get_data(self, data_store_name: str, **kwargs):
 
         Args:
             data_store_name (str): Name of data store.
-            **kwargs: queryparams (json/bson): JSON / BSON string.
+
+        Keyword Args:
+            **kwargs: Arbitrary keyword arguments.
+            queryparams (dict): dict following pymongo convention.
+            sort_query (List of tuples): (field_to_sort, sort_direction)
+                direction can be ASC/ASCENDING or DSC/DESCENDING.
```
[the remark this answers is not in the document]
Comment (akash2144) [reply common/db_store/db.py]: Changed to DESC.
DESC is abbreviated form of Descending.

File: common/db_store/db.py [hunk 3]
```diff
@@ -90,26 +90,66 @@ def get_data(self, data_store_name: str, **kwargs):
...
+            limit (int): Limit for no. of documents returned by query.
```
[the remark this answers is not in the document]
Comment (akash2144) [reply common/db_store/db.py]: To get the docs from 11 to 20 we need to set skip=10 & limit=20
Added Example in docstring.

File: common/db_store/db.py [hunk 4]
```diff
@@ -90,26 +90,66 @@ def get_data(self, data_store_name: str, **kwargs):
...
+            skip (int): To skip the first x records of the returned results.
```
[the remark this answers is not in the document]
Comment (akash2144) [reply common/db_store/db.py]: Added example.

File: common/db_store/db.py [hunk 5]
```diff
@@ -90,26 +90,66 @@ def get_data(self, data_store_name: str, **kwargs):
...
+        count = kwargs.get('count')
 
         try:
-            results = self._db[data_store_name].find(query_params)
+            if sort_query is not None:
+                sort_fields = []
+                for sort_field, sort_dir in sort_query:
+                    if sort_dir.upper() in ('ASC', 'ASCENDING'):
+                        sort_fields.append((sort_field, pymongo.ASCENDING))
+                    elif sort_dir.upper() in ('DSC', 'DESCENDING'):
```
[the remark this answers is not in the document]
Comment (akash2144) [reply common/db_store/db.py]: Changed to DESC.

File: common/db_store/db.py (hunk 1 above)
Comment (akash2144) [reply common/db_store/db.py:109]: The objective of get_data method is to get list of records, also by-default it return count as None if you haven't set count=True. That's why I kept it as second parameter.

File: common/db_store/db.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mukhtar-inamdar) [reply common/db_store/db.py]: ok

**Review by mukhtar-inamdar — APPROVED** [review]
Done

**Review by shaileshsvc — APPROVED** [review]
Get other comments marked resolved.

**Pull request merged**